feat(x/ecocredit): update batch proto to support dynamic minting

robert-zaremba opened the pull request:
## Description

Closes: #924


---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

robert-zaremba commented on proto/regen/ecocredit/v1/state.proto:
```diff
@@ -10,224 +10,232 @@ option go_package = "github.com/regen-network/regen-ledger/x/ecocredit/core";
+
+  // escrowed_amount is the amount of credits in the batch that have been
+  // locked up in escrow for use in the marketplace.
+  string escrowed_amount = 5;
```
auto formatted

ryanchristo replied:
Can we finish the proto lint pull request and make the formatting changes there (#721)?

Some of the files updated in this pull request are not related to the proposed changes (e.g. basket proto files). Also, the space before the colon does not appear to be a standard proto style guideline, is this something you added?

ryanchristo replied:
> there is no space before the colon

You are adding spaces before the colon. Maybe that's what you meant to say.

```
primary_key : {fields : "id", auto_increment : true}
```

If this is the format we want to use, we should make sure this is consistent for everyone pushing code (#721). This is new to everyone else, so would be nice to make sure we are all aligned and using the same style.

robert-zaremba commented on x/ecocredit/go.mod:
```diff
@@ -21,6 +21,7 @@ require (
 	github.com/tendermint/tendermint v0.34.15
 	github.com/tendermint/tm-db v0.6.7
 	github.com/thanhpk/randstr v1.0.4
+	golang.org/x/mod v0.4.2
```
no idea why this was added

clevinson commented:
Can we reduce this PR to just updated proto files (no codegen files, and no stubs implementing interfaces, etc.)? 

I think if its only a PR with proto files for now then it will be a great companion to the #924 issue so we can talk concretely about what an implementation hypothetically would look like.

robert-zaremba commented:
Reducing this PR only to proto updates.

robert-zaremba commented on proto/regen/ecocredit/v1/state.proto:
```diff
@@ -136,6 +136,11 @@ message BatchInfo {
+  // the future. Once the enable_future_minting is set to false, it can't be
+  // toggled any more.
+  bool enable_future_minting = 8;
 }
```
if we have `enable_future_minting`, `max_supply` is not needed any more. so I removed `max_supply`.

aaronc replied:
I actually think the default should be sealed. So if we want a different name let's use `open`

robert-zaremba replied:
It's in the state, so we control the defaults. Good naming is a challenge.
If we want to rename this, then let's use `open` everywhere.

robert-zaremba commented on proto/regen/ecocredit/v1/tx.proto:
```diff
@@ -165,6 +180,39 @@ message MsgCreateBatchResponse {
+    // the id of a transaction based on a type (tx id, serial number)
+    string id = 2;
+  }
+}
```
Both, the `note` and `origin_tx` will service the accounting purpose.

aaronc commented on proto/regen/ecocredit/v1/tx.proto:
```diff
@@ -23,6 +23,16 @@ service Msg {
   rpc CreateBatch(MsgCreateBatch) returns (MsgCreateBatchResponse);
 
+  // MintBatch issues new token in a given batch.
+  // Currently the issuer must be the Credit Class issuer, or an address
```
The issuer should be the issuer of the original batch. Not any credit class issuer

robert-zaremba replied:
done

aaronc commented on proto/regen/ecocredit/v1/tx.proto:
```diff
@@ -23,6 +23,16 @@ service Msg {
+  // Currently the issuer must be the Credit Class issuer, or an address
+  // autorized by him. The request will fail if the batch enable_future_minting
+  // was previously set to false.
+  // NOTE: this method is only for bridge purpose. It must not be used
```
```suggestion
  // NOTE: this method is only for bridging credits from other registries. It should not be used
```

aaronc commented on proto/regen/ecocredit/v1/tx.proto:
```diff
@@ -23,6 +23,16 @@ service Msg {
+  // was previously set to false.
+  // NOTE: this method is only for bridge purpose. It must not be used
+  // for issuing native credits on Regen. More specifically, we
+  // enable minting more credits in an existing batch, when the batch
```
```suggestion
  // enable minting more credits in an existing batch when the batch
```

aaronc commented on proto/regen/ecocredit/v1/tx.proto:
```diff
@@ -23,6 +23,16 @@ service Msg {
+  // NOTE: this method is only for bridge purpose. It must not be used
+  // for issuing native credits on Regen. More specifically, we
+  // enable minting more credits in an existing batch, when the batch
+  // represents a batch - vintage originally registered in other chain.
```
```suggestion
  // represents a batch - vintage originally registered an another chain or registry. 
```

aaronc reviewed (commented):
Generally looks good. I think minting permissions should be more restrictive

robert-zaremba commented:
+ changed `enable_future_minting` to `sealed`
+ updated docs
+ added `issuer` to BatchInfo

clevinson commented on proto/regen/ecocredit/v1/tx.proto:
```diff
@@ -165,6 +179,38 @@ message MsgCreateBatchResponse {
+  string issuer = 1;
+
+  // issuance are the credits issued in the batch.
+  repeated BatchIssuance issuance = 2;
```
If there are multiple `BatchIssuance`s for a given `MsgMintBatch` is it expected that they all have the same `origin_tx`?

We could alternatively evaluate nesting `origin_tx` inside BatchIssuance message, so that an on-chain minting event can batch together multiple issuances (coming from multiple different bridge transactions). This may be suboptimal though, as leveraging events becomes quite cumbersome and annoying if in the current event system if we have multiple event attributes corresponding to different origin bridge transactions for one SDK tx.

Also- do we need to support `origin_tx` information in `MsgCreateBatch()` in cases where a bridge action triggers a new batch creation? Or do we expect that the bridge creator does not mint any credits when they create a new batch for a bridge (so all minting of credits via a bridge always happen through a `MsgMintBatch()` call).

I see a few options:

### All issuances from bridges must happen through `MsgMintBatch()` 

In this case, we don't need `origin_tx` information in `MsgCreateBatch()`, and we should make sure we have some on-chain validation that `MsgCreateBatch()` fails if `sealed == false` and `issuance != nil`.

### Allow issuances from bridges in `MsgCreateBatch()`

In this case, we should add `origin_tx` into `MsgCreateBatch()`

robert-zaremba replied:
> Also- do we need to support origin_tx information in MsgCreateBatch() in cases where a bridge action triggers a new batch creation?

Good point. yes!

robert-zaremba replied:
I've updated the `MsgCreatBatch` - so bridge will create a batch and mint at the same time. It will be consistent with the minting we have now. Subsequent mints will use MsgMintBatch

robert-zaremba commented:
rename sealed -> unlocked

clevinson reviewed (commented):
Thanks for these updates @robert-zaremba !

A few suggestions in-line, and additionally I have one larger question about your proposal for utilizing events for accounting purposes.

In our current bridge strategy, the bridge operator is responsible (either manually or via an automated way) for initiating `MintBatch()` to issue newly bridged credits to a batch.  And this operator is the only person who can mint new credits to this batch.

Still, it would be nice to have some safety mechanisms that prevent the bridge operator from accidentally bridging credits twice.  If a bridge operator crashes, and is restarting and pulling a queue of bridge events on Polygon via querying EVM events, then how does the bridge operator ensure that it doesn't mint credits that it has already minted? Certainly it can query events for an `origin_tx` to ensure that a given Polygon Tx has not been bridged over yet, but what if it is communicating with a malicious API node? Or what if the full-node that the bridge is querying is just falling behind in processing blocks (which is quite often a possibility)?

I think it might be preferable to have some on-chain logic that enforces no duplicate mints get created for a given `origin_tx`. I know that it's been discussed in the past the possibility of having the state machine have access to events, but I also remember this being pushed back against by some folks on the SDK team -- so I'm not assuming that this will happen.

Given that, I think it would be more secure and preferable to store `origin_tx` in state (maybe as a simple field in `BatchInfo` (or a separate table entirely) called `origin_txs` that stores a list of all `origin_tx` fields as they get processed, so that when processing a new `MsgMintBatchRequiest()`, the state machine would verify that the given `origin_tx` has not been submitted yet.

clevinson commented on proto/regen/ecocredit/v1/tx.proto:
```diff
@@ -133,30 +143,34 @@ message MsgCreateBatch {
-  // BatchIssuance represents the issuance of some credits in a batch to a
-  // single recipient.
-  message BatchIssuance {
+  // If unlocked is set to true, we will seal the batch and disable the
```
```suggestion
  // If unlocked is set to false, we will seal the batch and disable the
```

clevinson commented on proto/regen/ecocredit/v1/state.proto:
```diff
@@ -136,6 +136,14 @@ message BatchInfo {
+
+  // sealed tells if it's possible to mint new credits in the future.
+  // Once the sealed is set to true, it can't be toggled any more.
+  bool sealed = 9;
```
Looks like this was already addressed. thx!

clevinson commented on proto/regen/ecocredit/v1/tx.proto:
```diff
@@ -23,6 +23,16 @@ service Msg {
+  // MintBatch issues new token in a given batch.
+  // Currently the issuer must be the Credit Class issuer, or an address
+  // autorized by him. The request will fail if the batch enable_future_minting
+  // was previously set to false.
```
```suggestion
  // autorized by him. The request will fail if the batch's `unlocked` field
  // is set to false.
```

aaronc commented:
I agree about origin tx uniqueness, but please let's use a join table rather than a list in keeping with 1NF

robert-zaremba commented on proto/regen/ecocredit/v1/tx.proto:
```diff
@@ -23,6 +23,16 @@ service Msg {
+  // for issuing native credits on Regen. More specifically, we
+  // enable minting more credits in an existing batch, when the batch
+  // represents a vintage originally registered in another chain.
+  rpc MintBatch(MsgMintBatch) returns (MsgMintBatchResponse);
```
yes, doesn't sound very English. How about `MintBatchCredits`?

robert-zaremba commented:
+ updated events (see comment in PR)
+ added orig tx index
+ added OrigTx and note to `MsgCreateBatch`
+ rename MintBatch to MintBatchCredits
+ use open instead of unlocked.

robert-zaremba commented on proto/regen/ecocredit/v1/types.proto:
```diff
@@ -34,3 +34,13 @@ message Params {
+  // Ethereum, Verra,
+  string typ = 1;
+  // the id of a transaction based on a type (tx id, serial number)
+  string id = 2;
```
Normally I would prefer to  not storing everything. We have events to keep that information. But events are not guaranteed to be permanent. 
So ideally we don't store everything here and we sort out the events in SDK.

What do you think?

robert-zaremba commented on proto/regen/ecocredit/v1/state.proto:
```diff
@@ -236,3 +244,24 @@ message BatchSupply {
+  string note = 3;
+
+  // batch_denom is the unique ID of the credit batch.
+  string batch_denom = 4;
```
Normally I would prefer to not storing everything. We have events to keep that information. But events are not guaranteed to be permanent.
So ideally we don't store everything here and we sort out the events in SDK.

What do you think?

robert-zaremba replied:
in fact, since I'm adding this fields here, I've slimmed down the `EventMintBatchCredits` event

ryanchristo commented on proto/regen/ecocredit/v1/tx.proto:
```diff
@@ -23,6 +24,17 @@ service Msg {
+  // for issuing native credits on Regen. More specifically, we
+  // enable minting more credits in an existing batch, when the batch
+  // represents a vintage originally registered in another chain.
+  rpc MintBatchCredits(MsgMintBatchCredits)
+      returns (MsgMintBatchCreditsResponse);
```
Can you comment this line out for the time being and then we can uncomment when implementing (#973). Otherwise anyone who makes proto changes will face errors because this message is not yet implemented.

ryanchristo replied:
```suggestion
  // rpc MintBatchCredits(MsgMintBatchCredits)
      returns (MsgMintBatchCreditsResponse);
```

ryanchristo commented on proto/regen/ecocredit/v1/tx.proto:
```diff
@@ -133,30 +145,43 @@ message MsgCreateBatch {
-    // effectively retired by the issuer on receipt. Decimal values are
-    // acceptable.
-    string retired_amount = 3;
+// BatchIssuance represents the issuance of some credits in a batch to a
+// single recipient.
+message BatchIssuance {
```
This will also break tests:

```
core/msg_create_batch_test.go:32:18: undefined: MsgCreateBatch_BatchIssuance
```

Can you update any references of `MsgCreateBatch_BatchIssuance` to `BatchIssuance` before merging?

ryanchristo commented:
When writing the proto API for new features, we should either include dummy implementation methods or comment out any new rpc methods that will cause tests to fail after generating the proto files.

Ideally changes to the existing API would not be made in the same pull request, which would prevent the need to update any code in the existing implementation. In cases like this where `BatchIssuance` is being pulled out of `BatchInfo`, we should generate the proto files and make sure the necessary changes are made in the existing implementation.

A potential process for designing new features starting with proto API updates:

1. a pull request is created with only proto API definitions
2. the pull request is reviewed and the design is approved
3. proto files are generated and, if necessary, minimal changes are made to the existing implementation, which will allow proto files to be generated in other pull requests without breaking tests

clevinson commented on proto/regen/ecocredit/v1/tx.proto:
```diff
@@ -165,6 +190,34 @@ message MsgCreateBatchResponse {
+  // issuance are the credits issued in the batch.
+  repeated BatchIssuance issuance = 3;
+
+  // open=true keeps the batch unlocked for future minting.
+  // Othrwise we will seal the batch after the mint.
+  // Throws an error if the batch is already sealed.
+  bool open = 4;
```
Could we update this to actually be a separate message? I think in most use cases (thinking of bridges), the issuer won't actually ever want to seal the batch.  Adding this here requires the user to explicitly set this value every time they call `MsgMintBatchCredits`, and if they forget, then it would lock the batch and prevent the issuer from being able to issue more credits to that vintage (assuming we have the appropriate restrictions on project <> credit type duplicate issuances).

I think I'd rather see something like:
```proto
message MsgSealBatch {
  string issuer = 1;
}
```

clevinson replied:
@aaronc I know you had initially proposed this `bool open = 4;` what do you think of this upgrade?

aaronc replied:
I don't think it makes that big of a difference. This should all be done automatically anyway. But I don't object

robert-zaremba replied:
Yesterday we had a session with Cory about the process and we found that it would make sense to separate the toggle functionality to a separate message. So if nobody objects I will make the update.

clevinson commented:
I think the proto gen stubs suggested from @ryanchristo makes sense. 

Im happy to give a concept ACK on the general design now- the only api change i'd still like to see is a MsgSealBatch message separated out. 

@robert-zaremba can you add go stubs and address the related go changes to prevent tests from breaking?

robert-zaremba commented:
Sorry but we are going back and forth without being constructive. 2 weeks ago I started with basic implementation for Msg methods and wanted to add dummy implementation for the RPC implementation while waiting for reviews (which would prevent troubles when someone generates the code and introduces additional changes). It got a criticism (https://github.com/regen-network/regen-ledger/pull/937#issuecomment-1079395808and and and discord) to ONLY do proto code (so NO code generation and NO other methods). 

The tests are not breaking because this PR doesn't include generated code.

I prefer to merge this PR and do other PR with implementation to not hold on this PR any more.

robert-zaremba commented:
I've added `MsgSealBatch`

clevinson commented:
Sorry about the back & forth @robert-zaremba, but I also remembered a bit more than dummy implementation in what you originally was putting forward (I think I remember some implementation work as well). Happy to have this merged now, and let's discuss maybe on one of our next dev calls to make sure we're not going back and forth too much with process around proto PRs as design.

glandua commented:
Great work Robert.  Although I was originally pretty restrained about the dynamic batch approach, I am quite pleased how this has turned out.

robert-zaremba commented:
@ryanchristo , @clevinson there was:

> basic implementation for Msg methods

robert-zaremba commented:
thanks for merging! :beers:
